tests/s3_data_plane_tests.c

```diff
@@ -3431,3 +3432,200 @@ static int s_test_s3_not_satisfiable_range(struct aws_allocator *allocator, void
+}
+
+static bool s_copy_test_completion_predicate(void *arg) {
+    struct copy_object_test_data *test_data = arg;
+    return test_data->execution_completed;
+}
+
+static int s_test_s3_copy_object_helper(
+    struct aws_allocator *allocator,
+    void *ctx,
+    struct aws_byte_cursor source_key,
+    struct aws_byte_cursor destination_key) {
+
+    (void)ctx;
+
+    struct aws_s3_tester tester;
+    AWS_ZERO_STRUCT(tester);
+    ASSERT_SUCCESS(aws_s3_tester_init(allocator, &tester));
+
+    struct aws_s3_client_config client_config;
+    AWS_ZERO_STRUCT(client_config);
+
+    ASSERT_SUCCESS(aws_s3_tester_bind_client(
+        &tester, &client_config, AWS_S3_TESTER_BIND_CLIENT_REGION | AWS_S3_TESTER_BIND_CLIENT_SIGNING));
+
+    struct aws_s3_client *client = aws_s3_client_new(allocator, &client_config);
+
+    struct aws_byte_cursor source_bucket = g_test_bucket_name;
+    struct aws_byte_cursor destination_bucket = g_test_bucket_name;
+
+    char endpoint[1024];
+    snprintf(
+        endpoint,
+        sizeof(endpoint),
+        "%.*s.s3.%s.amazonaws.com",
+        (int)destination_bucket.len,
+        destination_bucket.ptr,
+        g_test_s3_region.ptr);
+
+    /* creates a CopyObject request */
+    struct aws_http_message *message = copy_object_request_new(
+        allocator, source_bucket, source_key, aws_byte_cursor_from_c_str(endpoint), destination_key);
+
+    struct copy_object_test_data test_data;
+    AWS_ZERO_STRUCT(test_data);
+
+    test_data.c_var = (struct aws_condition_variable)AWS_CONDITION_VARIABLE_INIT;
+    aws_mutex_init(&test_data.mutex);
+
+    struct aws_s3_meta_request_options meta_request_options = {
+        .user_data = &test_data,
+        .body_callback = NULL,
+        .signing_config = client_config.signing_config,
+        .finish_callback = s_copy_object_meta_request_finish,
+        .headers_callback = NULL,
+        .message = message,
+        .shutdown_callback = NULL,
+        .type = AWS_S3_META_REQUEST_TYPE_COPY_OBJECT,
+    };
+
+    struct aws_s3_meta_request *meta_request = aws_s3_client_make_meta_request(client, &meta_request_options);
+    if (meta_request == NULL) {
+        printf("*** meta_request IS NULL\n");
+    }
+
+    /* wait completion of the meta request */
+    aws_mutex_lock(&test_data.mutex);
+    aws_condition_variable_wait_pred(&test_data.c_var, &test_data.mutex, s_copy_test_completion_predicate, &test_data);
+    aws_mutex_unlock(&test_data.mutex);
+
+    /* assert error_code and respose_status */
+    ASSERT_INT_EQUALS(AWS_ERROR_SUCCESS, test_data.meta_request_error_code);
+    ASSERT_INT_EQUALS(AWS_HTTP_STATUS_CODE_200_OK, test_data.response_status_code);
+
+    aws_s3_meta_request_release(meta_request);
+    aws_mutex_clean_up(&test_data.mutex);
+    aws_http_message_destroy(message);
+    aws_s3_client_release(client);
+    client = NULL;
+
+    aws_s3_tester_clean_up(&tester);
+
+    return 0;
+}
+
+AWS_TEST_CASE(test_s3_copy_small_object, s_test_s3_copy_small_object)
```

Comment:
it seems to me that the state machinery around which network path is taken had more variations than just small and large objects. Are you sure at least all the branches that can be taken without a proper wire mocking system are taken?

Reply:
Added one more test checking for failure due to invalid source key. The intermediary request failures are not being tested currently and will require a proper wire mocking system. I'll handle the mocking in a distinct PR.